app/controllers/container_topology_controller.rb

```diff
@@ -0,0 +1,105 @@
+class ContainerTopologyController < ApplicationController
+  before_action :check_privileges
+  before_action :get_session_data
+  after_action :cleanup_action
+  after_action :set_session_data
+
+  def show
+    topology = build_topology
+    @topologyitems = topology[:items].to_json
+    @topologyrelations = topology[:relations].to_json
+    @topologykinds = topology[:kinds].to_json
+  end
+
+  def index
+    redirect_to :action => 'show'
+  end
+
+  def data
+    render :json => {:data => build_topology}
+  end
+
+  private
+
+  def get_session_data
+    @layout = "container_topology"
+  end
+
+  def set_session_data
+    session[:layout] = @layout
+  end
+
+  def build_topology
+    nodes, services = entities
+    topology = {}
+    topo_items = {}
+    links = []
+
+    nodes.each do |n|
+      topo_items[n.ems_ref] = {:metadata => {:id  => n.ems_ref, :name => n.name}, :kind => "Node"}
+      if n.container_groups.size > 0
+        n.container_groups.each do |cg|
+          topo_items[cg.ems_ref] = build_entity(cg, "ContainerGroup")
+          links << build_link(n, cg)
+          cg.containers.each do |c|
+            topo_items[c.ems_ref] = build_entity(c, "Container")
+            links << build_link(cg, c)
+          end
+        end
+
+      end
+      unless n.lives_on.nil?
+        kind = n.lives_on.type.end_with?('Vm') ? "VM" : "Host"
```

Comment:
This might be bad...avoid String manipulation especially on the Type column.  Prefer `n.lives_on.kind_of?(Vm) ? "Vm" : "Host"`.  @matthewd's rewrite will completely screw up this code.

Or perhaps just use the dictionary or I18N stuff to convert the object you have into a String (i.e. `ui_lookup(:model => n.lives_on.class)`, or something like that...@h-kataria might be able to explain that better.)